src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+        {
+            _compilation = compilation;
+            _symbol = symbol;
+            _useUpdatedEscapeRules = useUpdatedEscapeRules;
+            _diagnostics = diagnostics;
+            _inUnsafeRegion = inUnsafeRegion;
+            _localScopeDepth = Binder.CurrentMethodScope - 1;
+            _localEscapeScopes = localEscapeScopes;
+        }
+
+        private ref struct LocalScope
+        {
+            private readonly RefSafetyAnalysis _analysis;
+            private readonly ImmutableArray<LocalSymbol> _locals;
+
+            public LocalScope(RefSafetyAnalysis analysis, ImmutableArray<LocalSymbol> locals)
+            {
+                _analysis = analysis;
+                _locals = locals;
+                _analysis._localScopeDepth++;
+                foreach (var local in locals)
+                {
+                    _analysis.AddLocalScopes(local, refEscapeScope: _analysis._localScopeDepth, valEscapeScope: Binder.CallingMethodScope);
+                }
+            }
+
+            public void Dispose()
+            {
+                foreach (var local in _locals)
+                {
+                    _analysis.RemoveLocalScopes(local);
+                }
+                _analysis._localScopeDepth--;
+            }
+        }
+
+        private ref struct UnsafeRegion
+        {
+            private readonly RefSafetyAnalysis _analysis;
+            private readonly bool _previousRegion;
+
+            public UnsafeRegion(RefSafetyAnalysis analysis, bool inUnsafeRegion)
+            {
+                _analysis = analysis;
+                _previousRegion = analysis._inUnsafeRegion;
+                _analysis._inUnsafeRegion = inUnsafeRegion;
+            }
+
+            public void Dispose()
+            {
+                _analysis._inUnsafeRegion = _previousRegion;
+            }
+        }
+
+        private ref struct PatternInput
+        {
+            private readonly RefSafetyAnalysis _analysis;
+            private readonly uint _previousInputValEscape;
+
+            public PatternInput(RefSafetyAnalysis analysis, uint patternInputValEscape)
+            {
+                _analysis = analysis;
+                _previousInputValEscape = analysis._patternInputValEscape;
+                _analysis._patternInputValEscape = patternInputValEscape;
+            }
+
+            public void Dispose()
+            {
+                _analysis._patternInputValEscape = _previousInputValEscape;
+            }
+        }
+
+        private ref struct PlaceholderRegion
+        {
+            private readonly RefSafetyAnalysis _analysis;
+            private readonly ArrayBuilder<BoundValuePlaceholderBase> _placeholders;
+
+            public PlaceholderRegion(RefSafetyAnalysis analysis, ArrayBuilder<(BoundValuePlaceholderBase, uint)> placeholders)
+            {
+                _analysis = analysis;
+                _placeholders = ArrayBuilder<BoundValuePlaceholderBase>.GetInstance(placeholders.Count);
```

Comment:
nit: Why build a fresh storage of placeholders, as opposed to just holding onto the `placeholders` `ArrayBuilder` that we have?